internal/revision/database/revision.go

```diff
@@ -0,0 +1,292 @@
+	if wrapperKey == "" {
+		return nil, fmt.Errorf("no KMS key ID passed in to revision.New")
+	}
+	if len(wrapperAAD) == 0 {
+		return nil, fmt.Errorf("no AAD provided")
+	}
+	if kms == nil {
+		return nil, fmt.Errorf("no KeyManager provided")
+	}
+	return &RevisionDB{
+		db:           db,
+		wrapperKeyID: wrapperKey,
+		wrapperAAD:   wrapperAAD,
+		keyManager:   kms,
+	}, nil
+}
+
+// DestroyKey zeros out the wrapped key and marks the key as allowed=false.
+func (rdb *RevisionDB) DestroyKey(ctx context.Context, keyID int64) error {
+	logger := logging.FromContext(ctx)
+	logger.Warnf("destroying key material for revision key ID %v", keyID)
+	zero := []byte{0}
+
+	if err := rdb.db.InTx(ctx, pgx.Serializable, func(tx pgx.Tx) error {
+		result, err := tx.Exec(ctx, `
+			UPDATE
+				RevisionKeys
+			SET
+				wrapped_cipher = $1, aad = $2, allowed = $3
+			WHERE
+				kid = $4
+		`, zero, zero, false, keyID)
+		if err != nil {
+			return fmt.Errorf("updating revisionkey: %w", err)
+		}
+		if result.RowsAffected() != 1 {
+			return fmt.Errorf("revision key was not updated as expected")
+		}
+		return nil
+	}); err != nil {
+		logger.Errorf("failed to destroy revision kid: %v: %v", keyID, err)
+	}
+	return nil
+}
+
+// GetAllowedRevisionKeyIDs returns just the IDs of still allowed keys.
+// Once the keys have been unwrapped, there is no reason to continue to unwrap them.
+func (rdb *RevisionDB) GetAllowedRevisionKeyIDs(ctx context.Context) (int64, map[int64]struct{}, error) {
+	var effectiveID int64
+	keys := make(map[int64]struct{})
+	if err := rdb.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				kid
+			FROM
+				revisionkeys
+			WHERE
+				allowed=$1
+			ORDER BY created_at DESC`, true)
+		if err != nil {
+			return nil
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			if err := rows.Err(); err != nil {
+				return fmt.Errorf("failed to iterate: %w", err)
+			}
+			var id int64
+			if err := rows.Scan(&id); err != nil {
+				return err
+			}
+			keys[id] = struct{}{}
+			if effectiveID == 0 {
+				effectiveID = id
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return 0, nil, fmt.Errorf("unable to read keys: %w1", err)
+	}
+	return effectiveID, keys, nil
+}
+
+// GetAllowedRevisionKeys returns all of the curently allowed revision keys.
+// This method will unwrap all of the keys so that they can be used to create and verify
+// revision tokens.
+func (rdb *RevisionDB) GetAllowedRevisionKeys(ctx context.Context) (int64, []*RevisionKey, error) {
+	logger := logging.FromContext(ctx)
+	var effectiveID int64
+	keys := make([]*RevisionKey, 0)
+	if err := rdb.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				kid, aad, wrapped_cipher, created_at, allowed
+			FROM
+				revisionkeys
+			WHERE
+				allowed=$1
+			ORDER BY created_at DESC`, true)
```

Comment:
This is probably worth commenting or highlighting somewhere - the []*RevisionKey will be ordered chronologically.